pkg/dedup/chunk_iter.go

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
```

Comment:
For other reviewers: Those things were moved, and (almost?) nothing was changed

Wonder if we can do the move in separate PR if we have time for easier review 🤔

Reply:
Almost the same as compact merger. But it supports downsampled encoding. Please review from the overlapMerger struct. Thanks

Reply:
I don't think GitHub supports it but locally you can do `git diff --color-moved main...`. It makes it clear what code has been simply moved :wink:

Reply:
The core part is https://github.com/thanos-io/thanos/pull/4239/files#diff-b17d8397d84790d382e887eae8d7718260a36c1ff75d71150e9306437a3f90c9R153-R357. PTAL. They are used to vertically merge series.